File: Data Structures & Algorithms/buy-and-sell-crypto/submission-1.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        // Time : O(n)
        // Space: O(1)

        // int left = 0;
        // int right = 1;

        // int maxP = 0;

        // while (right < prices.size()){
        //     int profit = prices[right] - prices[left];
        //     maxP = max(maxP, profit);

        //     if (prices[left] > prices[right]) left = right;
        //     right++;
        // }

        // return maxP;

        int maxP = 0;
        int buy = prices[0];

        for(int i = 1; i < prices.size(); i++){
            int profit = prices[i] - buy;

            maxP = max(maxP, profit);
            buy = min(buy, prices[i]);
        }
        return maxP;
    }
};

File: Data Structures & Algorithms/eating-bananas/submission-1.py
class Solution:
    def isPossible(self, piles: List[int], h: int, k: int) -> bool:
        print("k: ", k)
        time = 0
        for p in piles:
            time += math.ceil(p / k)

            if time > h: 
                print("returning false")
                return False

        print("returning true")
        return True

    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        # longest amount of time to eat pile
        max_time = max(piles)
        print("max_time: ", max_time)

        low, high = 1, max_time

        while low < high:
            mid = low + (high - low) // 2
            print("mid: ", mid)
            # search left half of solution space
            if self.isPossible(piles, h, mid):
                high = mid

            else:
                low = mid + 1
            print("low: ", low)
            print("high: ", high)

        return high
